pkg/updater/repository/repository.go

```diff
@@ -0,0 +1,324 @@
+	repository, err := openRepository(r.RootPath)
+	if err != nil {
+		return nil, err
+	}
+	return &Status{
+		Stable:     repository.stable.Target(),
+		Experiment: repository.experiment.Target(),
+	}, nil
+}
+
+// Create creates a fresh new repository at the given root path
+// and moves the given stable source path to the repository as the first stable.
+// If a repository already exists at the given path, it is fully removed.
+//
+// 1. Remove the previous repository if it exists.
+// 2. Create the root directory.
+// 3. Move the stable source to the repository.
+// 4. Create the stable link.
+func (r *Repository) Create(name string, stableSourcePath string) error {
+	err := os.RemoveAll(r.RootPath)
+	if err != nil {
+		return fmt.Errorf("could not remove previous repository: %w", err)
+	}
+	err = os.MkdirAll(r.RootPath, 0755)
+	if err != nil {
+		return fmt.Errorf("could not create packages root directory: %w", err)
+	}
+	repository, err := openRepository(r.RootPath)
+	if err != nil {
+		return err
+	}
+	err = repository.setStable(name, stableSourcePath)
+	if err != nil {
+		return fmt.Errorf("could not set first stable: %w", err)
+	}
+	return nil
+}
+
+// SetExperiment moves the given source path to the repository and sets it as the experiment.
+//
+// 1. Cleanup the repository.
+// 2. Move the experiment source to the repository.
+// 3. Set the experiment link to the experiment package.
+func (r *Repository) SetExperiment(name string, sourcePath string) error {
+	repository, err := openRepository(r.RootPath)
+	if err != nil {
+		return err
+	}
+	err = repository.cleanup()
+	if err != nil {
+		return fmt.Errorf("could not cleanup repository: %w", err)
+	}
+	if !repository.stable.Exists() {
+		return fmt.Errorf("stable package does not exist, invalid state")
+	}
+	err = repository.setExperiment(name, sourcePath)
+	if err != nil {
+		return fmt.Errorf("could not set experiment: %w", err)
+	}
+	return nil
+}
+
+// PromoteExperiment promotes the experiment to stable.
+//
+// 1. Cleanup the repository.
+// 2. Set the stable link to the experiment package.
+// 3. Delete the experiment link.
+// 4. Cleanup the repository to remove the previous stable package.
+func (r *Repository) PromoteExperiment() error {
+	repository, err := openRepository(r.RootPath)
+	if err != nil {
+		return err
+	}
+	err = repository.cleanup()
+	if err != nil {
+		return fmt.Errorf("could not cleanup repository: %w", err)
+	}
+	if !repository.stable.Exists() {
+		return fmt.Errorf("stable package does not exist, invalid state")
+	}
+	if !repository.experiment.Exists() {
+		return fmt.Errorf("experiment package does not exist, invalid state")
+	}
+	err = repository.stable.Set(*repository.experiment.packagePath)
+	if err != nil {
+		return fmt.Errorf("could not set stable: %w", err)
+	}
+	err = repository.experiment.Delete()
+	if err != nil {
+		return fmt.Errorf("could not delete experiment link: %w", err)
+	}
+	err = repository.cleanup()
```

Comment:
Do we need to do another cleanup here?

Reply:
Yes to free up the space after the delete as soon as possible.